File: Algorithms/Easy/2215.Find-The-Difference-Of-Two-Arrays/FindTheDifferenceOfTwoArrays.go
package easy

// Time complexity: O(n)
// Space complexity: O(n)
func findDifference(nums1 []int, nums2 []int) [][]int {
	answer := make([][]int, 2)
	nums1set := arrayToSet(nums1)
	nums2set := arrayToSet(nums2)

	for num := range nums1set {
		if _, ok := nums2set[num]; !ok {
			answer[0] = append(answer[0], num)
		}
	}

	for num := range nums2set {
		if _, ok := nums1set[num]; !ok {
			answer[1] = append(answer[1], num)
		}
	}

	return answer
}

func arrayToSet(array []int) map[int]struct{} {
	set := make(map[int]struct{})
	for _, num := range array {
		set[num] = struct{}{}
	}

	return set
}

File: Algorithms/Easy/2215.Find-The-Difference-Of-Two-Arrays/FindTheDifferenceOfTwoArrays.py
# Time complexity: O(n + m)
# Space complexity: O(max(n, m))
class Solution:
    def findDifference(self, nums1: list[int], nums2: list[int]) -> list[list[int]]:
        nums1_set = set(nums1)
        nums2_set = set(nums2)
        diffs = [[], []]

        for num in nums1_set:
            if num not in nums2_set:
                diffs[0].append(num)

        for num in nums2_set:
            if num not in nums1_set:
                diffs[1].append(num)

        return diffs
